packages/app/src/cli/services/dev/processes/dev-session.test.ts

```diff
@@ -0,0 +1,157 @@
+import {setupDevSessionProcess, pushUpdatesForDevSession} from './dev-session.js'
+import {DeveloperPlatformClient} from '../../../utilities/developer-platform-client.js'
+import {AppLinkedInterface} from '../../../models/app/app.js'
+import {AppEventWatcher} from '../app-events/app-event-watcher.js'
+import {buildAppURLForWeb} from '../../../utilities/app/app-url.js'
+import {testAppLinked, testDeveloperPlatformClient, testWebhookExtensions} from '../../../models/app/app.test-data.js'
+import {formData} from '@shopify/cli-kit/node/http'
+import {describe, expect, test, vi, beforeEach} from 'vitest'
+import {AbortSignal} from '@shopify/cli-kit/node/abort'
+import {flushPromises} from '@shopify/cli-kit/node/promises'
+
+vi.mock('@shopify/cli-kit/node/fs')
+vi.mock('@shopify/cli-kit/node/archiver')
+vi.mock('@shopify/cli-kit/node/http')
+vi.mock('../../../utilities/app/app-url.js')
+vi.mock('node-fetch')
+
+describe('setupDevSessionProcess', () => {
+  test('returns a dev session process with correct configuration', async () => {
+    // Given
+    const options = {
+      app: {} as AppLinkedInterface,
+      apiKey: 'test-api-key',
+      developerPlatformClient: {} as DeveloperPlatformClient,
+      storeFqdn: 'test.myshopify.com',
+      url: 'https://test.dev',
+      organizationId: 'org123',
+      appId: 'app123',
+      appWatcher: {} as AppEventWatcher,
+    }
+
+    // When
+    const process = await setupDevSessionProcess(options)
+
+    // Then
+    expect(process).toEqual({
+      type: 'dev-session',
+      prefix: 'dev-session',
+      function: pushUpdatesForDevSession,
+      options: {
+        app: options.app,
+        apiKey: options.apiKey,
+        developerPlatformClient: options.developerPlatformClient,
+        storeFqdn: options.storeFqdn,
+        url: options.url,
+        organizationId: options.organizationId,
+        appId: options.appId,
+        appWatcher: options.appWatcher,
+      },
+    })
+  })
+})
+
+describe('pushUpdatesForDevSession', () => {
+  let stdout: any
+  let stderr: any
+  let options: any
+  let developerPlatformClient: any
+  let appWatcher: AppEventWatcher
+  let app: AppLinkedInterface
+
+  beforeEach(() => {
+    vi.mocked(formData).mockReturnValue({append: vi.fn(), getHeaders: vi.fn()} as any)
+    stdout = {write: vi.fn()}
+    stderr = {write: vi.fn()}
+    developerPlatformClient = testDeveloperPlatformClient()
+    app = testAppLinked()
+    appWatcher = new AppEventWatcher(app)
+
+    options = {
+      developerPlatformClient,
+      appWatcher,
+      storeFqdn: 'test.myshopify.com',
+      appId: 'app123',
+      organizationId: 'org123',
+    }
+  })
+
+  test('creates a new dev session successfully when receiving the app watcher start event', async () => {
+    // When
+    await pushUpdatesForDevSession({stderr, stdout, abortSignal: new AbortSignal()}, options)
+    await appWatcher.start()
+    await flushPromises()
```

Comment:
This is optional because I'm not sure how practical it is. Am wondering about how to make the watcher easier to test.

Lots of these tests are running `.start()` and then waiting for flushPromises. I wonder if `start` could return something the caller could choose to await that resolves once the watcher has reached the settled state?

```
async function start(...) {
  const watchedFinishedStartup = new Promise((resolve) => {
    //...
    resolve()
  }
  return {
    watcherFinishedStartup
  }
}
```

Reply:
I tried this, but we still need the `flushPromises()`, the thing is that after "start" then the dev-session receives the event and start processing it, but we can await that, so the test continues to the `expects` before the event is processed and fails. 

I don't think there is an easy alternative to this 🤔